File: leetcode/golang/40_combination_sum_ii.go
func combinationSum2(candidates []int, target int) [][]int {
    var res [][]int
    var state []int
    var dfs func(candidates []int, target int, start int)
    dfs = func(candidates []int, target int, start int) {
        if target == 0 {
            res = append(res, append([]int(nil), state...))
            return
        }
        for i := start; i < len(candidates); i++ {
            if target >= candidates[i] {
               if i > start && candidates[i] == candidates[i-1] {
                   continue
               }
                state = append(state, candidates[i])
                dfs(candidates, target - candidates[i], i + 1)
                state = state[:len(state) - 1]
            }
        }
    }
    sort.Ints(candidates)
    dfs(candidates, target, 0)
    return res
}

File: leetcode/c++/40_combination_sum_ii.cpp
// https://leetcode.cn/problems/combination-sum-ii/description

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        dfs(candidates, target, 0);
        return res;
    }
private:
    vector<vector<int>> res;
    vector<int> state;
    void dfs(vector<int>& candidates, int target, int start) {
        if (target == 0) {
            res.push_back(state);
            return;
        }
        for (int i = start; i < candidates.size(); i++) {
            if (target >= candidates[i]) {
                if (i > start && candidates[i] == candidates[i-1]) 
                    continue;
                state.push_back(candidates[i]);
                dfs(candidates, target-candidates[i], i + 1);
                state.pop_back();
            }
        }
    }
};